--- lisp/org.el.orig	2010-03-13 00:40:33.798985209 +0100
+++ lisp/org.el	2010-03-14 13:18:11.326468951 +0100
@@ -4948,8 +4948,8 @@
 	   '(org-font-lock-hook)
 	   ;; Headlines
 	   `(,(if org-fontify-whole-heading-line
-		  "^\\(\\**\\)\\(\\* \\)\\(.*\n?\\)"
-		"^\\(\\**\\)\\(\\* \\)\\(.*\\)")
+	   	  "^\\(\\**\\)\\(\\* \\)\\(.*\n?\\)"
+	   	"^\\(\\**\\)\\(\\* \\)\\(.*\\)")
 	     (1 (org-get-level-face 1))
 	     (2 (org-get-level-face 2))
 	     (3 (org-get-level-face 3)))
@@ -4962,8 +4962,9 @@
 	   '("^[ \t]*|\\( *\\([$!_^/]\\) *|.*\\)|" (1 'org-formula t))
 	   '("| *\\(<[lr]?[0-9]*>\\)" (1 'org-formula t))
 	   ;; Drawers
-	   (list org-drawer-regexp '(0 'org-special-keyword t))
-	   (list "^[ \t]*:END:" '(0 'org-special-keyword t))
+	   ;(list org-drawer-regexp '(0 'org-special-keyword t))
+	   ;(list "^[ \t]*:END:" '(0 'org-special-keyword t))
+	   '(org-font-lock-add-drawer-faces)
 	   ;; Properties
 	   (list org-property-re
 		 '(1 'org-special-keyword t)
@@ -4979,25 +4980,18 @@
 				     (2 'org-footnote t)))
 	   '("^&?%%(.*\\|<%%([^>\n]*?>" (0 'org-sexp-date t))
 	   '(org-hide-wide-columns (0 nil append))
-	   ;; TODO lines
-	   (list (concat "^\\*+[ \t]+" org-todo-regexp "\\([ \t]\\|$\\)")
-		 '(1 (org-get-todo-face 1) t))
-	   ;; DONE
-	   (if org-fontify-done-headline
-	       (list (concat "^[*]+ +\\<\\("
-			     (mapconcat 'regexp-quote org-done-keywords "\\|")
-			     "\\)\\(.*\\)")
-		     '(2 'org-headline-done t))
-	     nil)
+	   ;; TODO states
+	   '(org-font-lock-add-todo-state-faces)
 	   ;; Priorities
 	   '(org-font-lock-add-priority-faces)
 	   ;; Tags
 	   '(org-font-lock-add-tag-faces)
 	   ;; Special keywords
-	   (list (concat "\\<" org-deadline-string) '(0 'org-special-keyword t))
-	   (list (concat "\\<" org-scheduled-string) '(0 'org-special-keyword t))
-	   (list (concat "\\<" org-closed-string) '(0 'org-special-keyword t))
-	   (list (concat "\\<" org-clock-string) '(0 'org-special-keyword t))
+	   '(org-font-lock-add-special-keyword-faces)
+	   ;(list (concat "\\<" org-deadline-string) '(0 'org-special-keyword t))
+	   ;(list (concat "\\<" org-scheduled-string) '(0 'org-special-keyword t))
+	   ;(list (concat "\\<" org-closed-string) '(0 'org-special-keyword t))
+	   ;(list (concat "\\<" org-clock-string) '(0 'org-special-keyword t))
 	   ;; Emphasis
 	   (if em
                (if (featurep 'xemacs)
@@ -5083,6 +5077,52 @@
 				 'font-lock-fontified t))
       (backward-char 1))))
 
+(defun org-font-lock-add-todo-state-faces (limit)
+  "Add the todo state faces."
+  (while (re-search-forward (concat "^\\*+[ \t]+" org-todo-regexp "\\(.*\\|$\\)") limit t)
+    (progn
+      ;; TODO lines
+      (add-text-properties
+       (match-beginning 1) (match-end 1) (list 'face (org-get-todo-face 1) 'font-lock-fontified t))
+      ;; DONE
+      ;; This part is buggy (sorry)
+      (if org-fontify-done-headline
+	  (save-excursion 
+	    (when (re-search-backward (concat "[ \t]+\\<\\(" 
+					      (mapconcat 'regexp-quote org-done-keywords "\\|")
+					      "\\)\\(.*\\)")
+				      (match-end 1) t)
+	      (add-text-properties
+	       (match-beginning 1) (match-end 1) (list 'face 'org-headline-done 'font-lock-fontified t))))
+	nil))))
+       
+(defun org-font-lock-add-special-keyword-faces (limit)
+  (progn
+    (save-excursion (while (re-search-forward (concat "\\<" org-deadline-string) limit t)
+		      (add-text-properties (match-beginning 0) (match-end 0) 
+					   (list 'face 'org-special-keyword 'font-lock-fontified t))))
+    (save-excursion (while (re-search-forward (concat "\\<" org-scheduled-string) limit t)
+		      (add-text-properties (match-beginning 0) (match-end 0) 
+					   (list 'face 'org-special-keyword 'font-lock-fontified t))))
+    (save-excursion (while (re-search-forward (concat "\\<" org-closed-string) limit t)
+		      (add-text-properties (match-beginning 0) (match-end 0) 
+					   (list 'face 'org-special-keyword 'font-lock-fontified t))))
+    (save-excursion (while (re-search-forward (concat "\\<" org-clock-string) limit t)
+		      (add-text-properties (match-beginning 0) (match-end 0) 
+					   (list 'face 'org-special-keyword 'font-lock-fontified t))))
+))
+
+(defun org-font-lock-add-drawer-faces (limit)
+  "Add the drawer faces."
+  (progn
+    (save-excursion
+      (while (re-search-forward org-drawer-regexp limit t)
+	(add-text-properties 
+	 (match-beginning 0) (match-end 0) (list 'face 'org-special-keyword 'font-lock-fontified t))))
+    (while (re-search-forward "^[ \t]*:END:" limit t)
+	(add-text-properties 
+	 (match-beginning 0) (match-end 0) (list 'face 'org-special-keyword 'font-lock-fontified t)))))
+
 (defun org-font-lock-add-priority-faces (limit)
   "Add the special priority faces."
   (while (re-search-forward "\\[#\\([A-Z0-9]\\)\\]" limit t)
@@ -6195,6 +6235,7 @@
 	     (t
 	      (or split (end-of-line 1))
 	      (newline (if blank 2 1)))))))
+	(remove-text-properties 0 (length head) '(display) head)
 	(insert head) (just-one-space)
 	(setq pos (point))
 	(end-of-line 1)
